common/gce.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Module for using the Google Compute Engine (GCE) API."""
+import threading
+
+import dateutil.parser
+
+from googleapiclient import discovery
+from oauth2client.client import GoogleCredentials
+
+thread_local = threading.local()  # pylint: disable=invalid-name
+
+
+def initialize():
+    """Initialize the thread-local configuration with things we need to use the
+    GCE API."""
+    thread_local.credentials = GoogleCredentials.get_application_default()
+    thread_local.service = discovery.build('compute',
+                                           'v1',
+                                           credentials=thread_local.credentials)
+
+
+def get_operations(project, zone):
+    """Generator that yields GCE operations for compute engine |project| and
+    |zone| in descendending order by time."""
+    zone_operations = thread_local.service.zoneOperations()  # pylint: disable=no-member
+    request = zone_operations.list(project=project,
+                                   zone=zone,
+                                   orderBy='creationTimestamp desc')
+    while request is not None:
+        response = request.execute()
+        for operation in response['items']:
+            yield operation
+
+        request = zone_operations.list_next(previous_request=request,
+                                            previous_response=response)
+
+
+def get_preemption_operations(operations):
```

Comment:
Where are these functions used ? Is some code missing from this CL ?

Reply:
They are in scheduler.py. One of the many great features of github reviews is hiding files that have the most changes (aka the most important files in a PR). :-(